src/data_readers/data_reader_python.cpp

```diff
@@ -229,32 +255,69 @@ int python_reader::get_linearized_label_size() const {
   return get_num_labels();
 }
 
-bool python_reader::fetch_datum(CPUMat& X, int data_id, int col) {
+bool python_reader::fetch_data_block(CPUMat& X,
+                                     El::Int thread_id,
+                                     El::Int mb_size,
+                                     El::Matrix<El::Int>& indices_fetched) {
 
-  // Lock mutex for the scope of this function
+  // Acquire Python GIL on first IO thread
+  // Note: Do nothing on other IO threads.
+  if (thread_id != 0) { return true; }
   auto& manager = python::manager::get_instance();
-  const auto lock = manager.get_mutex_guard();
+  python::global_interpreter_lock gil(manager);
+
+  // Get sample indices
+  python::object indices = PyList_New(0);
+  for (El::Int i = 0; i < mb_size; ++i) {
+    El::Int index = m_shuffled_indices[m_current_pos + i * m_sample_stride];
+    PyList_Append(indices, python::object(index));
+    indices_fetched.Set(i, 0, index);
+  }
 
-  // Get sample with Python
-  python::object args = Py_BuildValue("(i)", data_id);
-  python::object sample = PyObject_CallObject(m_sample_function, args);
-  sample = PyObject_GetIter(sample);
+  // Get samples using Python process pool
+  python::object samples = PyObject_CallMethod(m_process_pool,
+                                               "map",
+                                               "(O,O)",
+                                               m_sample_function.get(),
+                                               indices.get());
 
-  // Extract sample entries from Python iterator
+  // Extract sample entries from Python objects
   const El::Int sample_size = get_linearized_data_size();
-  for (El::Int row = 0; row < sample_size; ++row) {
-    python::object val = PyIter_Next(sample);
-    X(row, col) = PyFloat_AsDouble(val);
+  samples = PyObject_GetIter(samples);
+  for (El::Int col = 0; col < mb_size; ++col) {
+    python::object sample = PyIter_Next(samples);
+    sample = PyObject_GetIter(sample);
+    for (El::Int row = 0; row < sample_size; ++row) {
+      python::object val = PyIter_Next(sample);
+      X(row, col) = PyFloat_AsDouble(val);
+    }
   }
-  if (PyErr_Occurred()) { LBANN_ERROR("Python error detected"); }
 
   return true;
 }
 
+bool python_reader::fetch_datum(CPUMat& X, int data_id, int col) {
+  return true;
```

Comment:
It might be better to leave the default implementation (throws an error) than this, in case something changes and this gets called directly.